로그인 구현

vectorch9 opened the pull request:
- [x] security 설정 (cors, jwt)
- [x] 리프레시토큰 저장소(DB) 구현
- [x] 로그인 구현

File changed가 너무 길어질 것같아 한번 merge한 후 회원가입과 함께 `Response`, `Exception`부분을 리팩토링하려고 합니다.
따라서, `Response`와 `Exception`을 제외한 부분에 대해 의견을 남겨주시면 수정하겠습니다 !

Resolves: #16

vectorch9 committed:
etc: gradle build파일 jwt 의존성 수정

jwt-api뿐아닌 jwt-jackson, jwt-impl도 의존성에 추가

vectorch9 committed:
Fix: MemberRole 추가

MemberRole enum 구현
Member 엔티티에 MemberRole 컬럼추가

vectorch9 committed:
Feat: Jwt 관련 DTO들 구현

JwtDto, LoginRequest, RegisterRquest, ReissueRequest 구현
RefreshToken 저장을 위한 엔티티와 repository 구현

vectorch9 committed:
Feat: 로그인 구현

Jwt를 이용한 로그인 구현
아직 Response, Exception을 통일시키지 않았으므로 수정할 예정

seonpilKim commented on src/main/java/cloneproject/Instagram/config/JwtFilter.java:
```diff
@@ -0,0 +1,48 @@
+        }
+
+        filterChain.doFilter(request, response);
+    }
```
### 401, 403 예외 처리 커스터마이징
- 해당 예외는 시큐리티를 통해서 필터링 되기 때문에, Advice로 처리가 따로 불가하므로, `doFilterInternal` 메소드에서 try-catch 구문으로 각 예외마다 `request.setAttribute("errorCode", ErrorCode.xxxx)`으로 예외 코드를 설정해준 뒤, `JwtAccessDeniedHandler`와 `JwtAuthenticationEntryPoint`에서 처리해주시면 될 것 같습니다!
> 참고 포스팅: https://sas-study.tistory.com/362

seonpilKim commented on src/main/java/cloneproject/Instagram/config/WebSecurityConfig.java:
```diff
@@ -7,35 +7,73 @@
+
+        return source;
+    }
+
```
### CORS
![image](https://user-images.githubusercontent.com/68049320/141228610-bcd8c078-d58c-4e34-8534-bebcb70a8cd0.png)
- 위와 같이, `addAllowedOrigin` 메소드를 사용해서 `Access-Control-Allow-Origin`에 `*`을 넣는 방식은, `Credentials`을 `true`로 설정하는 것을 지원해주지 않습니다.
- 그러나 `allowedOriginPatterns` 메소드를 사용하시면 위의 문제가 해결됩니다!

seonpilKim commented on src/main/java/cloneproject/Instagram/controller/MemberController.java:
```diff
@@ -27,4 +33,27 @@
+        JwtDto result = memberService.reisuue(reissueRequest);
+        
+        return ResponseEntity.ok(result);
+    }
```
`@Validated`가 빠졌네요ㅎㅎ

seonpilKim commented on src/main/java/cloneproject/Instagram/service/JwtUserDetailsService.java:
```diff
@@ -0,0 +1,43 @@
+    private final MemberRepository memberRepository;
+
+    @Override
+    @Transactional
```
단순 조회 쿼리만 사용할 때는 `@Transactional(readonly = true)`으로 사용하면 성능이 조금 더 향상된다고 알고 있습니다!
> 참고 포스팅: https://willseungh0.tistory.com/75

seonpilKim commented on src/main/java/cloneproject/Instagram/entity/auth/RefreshToken.java:
```diff
@@ -0,0 +1,49 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @Column(name = "member_id", unique = true)
+    private String memberId;
```
members 테이블이랑 `@OneToOne` 매핑해두는 게 좋지 않을까요?
-> ✅members 테이블에 포함시키는 방식으로 변경

vectorch9 committed:
fix: RefreshToken 저장 방식변경

RefreshToken 저장방식 변경
 개별의 RefreshToken 저장소 사용 -> Member 엔티티 안에 포함시킴
시큐리티 설정 cors관련 부분 수정
@Validated 추가
Transantional의 import가 잘못된 부분 수정, readOnly사용 가능한 메서드는 추가
